### PR TITLE
Refactor the const eval diagnostic API

### PR DESCRIPTION
**Pull request by oli-obk**
* no longer report "const eval error" for things that have typeck errors
* errors and lints have saner spans and messages
* unified the diagnostic logic (const eval errors were slightly different depending on where they were reported, and there was also code duplication between the different reporters)
* report errors if an erroneous constant is used inside a promoted (fixes most of #50814)

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/middle/const_val.rs [hunk 1]
```diff
@@ -28,30 +30,29 @@ pub enum ConstVal<'tcx> {
...
 }
 
-#[derive(Clone, Debug)]
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
 pub enum ErrKind<'tcx> {
 
     NonConstPath,
-    UnimplementedConstVal(&'static str),
+    CouldNotResolve,
     IndexOutOfBounds { len: u64, index: u64 },
```
Comment (eddyb) [inline src/librustc/middle/const_val.rs]: Is this still needed? I think `Miri` covers it already.

**Comment by oli-obk** [comment]
@bors r=nikomatsakis

These keep dropping like flies. We should be able to remove const_val::ErrKind soonish

**Comment by oli-obk** [comment]
@bors r-

needs a rebase

File: src/librustc/traits/fulfill.rs [hunk 2]
```diff
@@ -534,8 +534,7 @@ fn process_predicate<'a, 'gcx, 'tcx>(
                             } else {
                                 Err(CodeSelectionError(ConstEvalFailure(ConstEvalErr {
                                     span: obligation.cause.span,
-                                    kind: ErrKind::UnimplementedConstVal("could not resolve")
-                                        .into(),
+                                    kind: ErrKind::CouldNotResolve.into(),
```
Comment (eddyb) [inline src/librustc/traits/fulfill.rs]: Wait, what error does miri produce from its own failed `Instance::resolve`?

Comment (oli-obk) [reply src/librustc/traits/fulfill.rs]: TypeckError: https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/eval_context.rs#L283

Comment (eddyb) [reply src/librustc/traits/fulfill.rs]: That's... disappointing. I believe failing to resolve means "still too polymorphic", does miri not have any general errors for that, short of the layout ones?

Comment (oli-obk) [reply src/librustc/traits/fulfill.rs]: No, but I will add one when refactoring away the old ctfe error enum

**Pull request merged**